driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -0,0 +1,524 @@
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.async.AsyncAggregateResponseBatchCursor;
+import com.mongodb.internal.async.SingleResultCallback;
+import com.mongodb.internal.async.function.AsyncCallbackSupplier;
+import com.mongodb.internal.binding.AsyncConnectionSource;
+import com.mongodb.internal.connection.AsyncConnection;
+import com.mongodb.internal.connection.Connection;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonInt64;
+import org.bson.BsonString;
+import org.bson.BsonTimestamp;
+import org.bson.BsonValue;
+import org.bson.codecs.BsonDocumentCodec;
+import org.bson.codecs.Decoder;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.assertNull;
+import static com.mongodb.assertions.Assertions.assertTrue;
+import static com.mongodb.assertions.Assertions.fail;
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.FIRST_BATCH;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.MESSAGE_IF_CLOSED_AS_CURSOR;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.MESSAGE_IF_CONCURRENT_OPERATION;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.NEXT_BATCH;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.NO_OP_FIELD_NAME_VALIDATOR;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.getMoreCommandDocument;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.translateCommandException;
+import static com.mongodb.internal.operation.OperationHelper.LOGGER;
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+
+class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T> {
+
+    private final MongoNamespace namespace;
+    private final int limit;
+    private final Decoder<T> decoder;
+    private final long maxTimeMS;
+    @Nullable
+    private final BsonValue comment;
+
+    private final AtomicInteger count = new AtomicInteger();
+    private final AtomicBoolean processedInitial = new AtomicBoolean();
+    private final AtomicReference<CommandCursorResult<T>> commandCursorResult = new AtomicReference<>();
+    private final int maxWireVersion;
+    private final boolean firstBatchEmpty;
+    private final ResourceManager resourceManager;
+    private int batchSize;
+
+    AsyncCommandBatchCursor(
```

Comment:
It could be made the same, however, there is no async Iterator API (which is why Publishers / Flows were invented and subsequently added to the jdk).

For async the API would be very heavy / require nested callbacks for use:

```
hasNext(SingleResultCallback<Boolean>)
next(SingleResultCallback<List<T>>)
```

Which can be replaced by just:

```
next(SingleResultCallback<List<T>>)
isClosed()
```

Mongodb cursors are self describing and have reached the end of their results.

The other main difference it the while loop used for getMores works for sync but in async it becomes a nested callback loop.

Reply:
That said there is no reason `hasNext()` could be added instead of `isClosed()`. I'll take a look and try to increase the code reuse.

Reply:
I opted to leave as is, as the only consumer is internal (BatchCursorFlux) it has a check in recurse cursor:

```
if (batchCursor.isClosed()) {
  sink.complete();
} else {
  // Fetches more results.
```